crates/primitives/src/filter.rs

```diff
@@ -0,0 +1,1214 @@
+         let mut bytes = [0; 32];
+         src.to_big_endian(&mut bytes);
+         ValueOrArray::Value(Some(H256::from(bytes)))
+     }
+ }
+
+ impl<T> Serialize for ValueOrArray<T>
+ where
+     T: Serialize,
+ {
+     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+     where
+         S: Serializer,
+     {
+         match self {
+             ValueOrArray::Value(inner) => inner.serialize(serializer),
+             ValueOrArray::Array(inner) => inner.serialize(serializer),
+         }
+     }
+ }
+
+ impl<'a, T> Deserialize<'a> for ValueOrArray<T>
+ where
+     T: DeserializeOwned,
+ {
+     fn deserialize<D>(deserializer: D) -> Result<ValueOrArray<T>, D::Error>
+     where
+         D: Deserializer<'a>,
+     {
+         let value = serde_json::Value::deserialize(deserializer)?;
+
+         if value.is_null() {
+             return Ok(ValueOrArray::Array(Vec::new()));
+         }
+
+         #[derive(Deserialize)]
+         #[serde(untagged)]
+         enum Variadic<T> {
+             Value(T),
+             Array(Vec<T>),
+         }
+
+         match serde_json::from_value::<Variadic<T>>(value).map_err(|err| {
+             serde::de::Error::custom(format!("Invalid variadic value or array type: {err}"))
+         })? {
+             Variadic::Value(val) => Ok(ValueOrArray::Value(val)),
+             Variadic::Array(arr) => Ok(ValueOrArray::Array(arr)),
+         }
+     }
+ }
+
+ /// Support for matching [Filter]s
+ #[derive(Debug, Default)]
+ pub struct FilteredParams {
+     pub filter: Option<Filter>,
+     pub flat_topics: Vec<ValueOrArray<Option<H256>>>,
+ }
+
+ impl FilteredParams {
+     pub fn new(filter: Option<Filter>) -> Self {
+         if let Some(filter) = filter {
+             let flat_topics = filter.flatten();
+             FilteredParams { filter: Some(filter), flat_topics }
+         } else {
+             Default::default()
+         }
+     }
+
+     /// Returns the [BloomFilter] for the given address
+     pub fn address_filter(address: &Option<ValueOrArray<Address>>) -> BloomFilter {
+         address.as_ref().map(address_to_bloom_filter).unwrap_or_default()
+     }
+
+     /// Returns the [BloomFilter] for the given topics
+     pub fn topics_filter(topics: &Option<Vec<ValueOrArray<Option<H256>>>>) -> Vec<BloomFilter> {
+         let mut output = Vec::new();
+         if let Some(topics) = topics {
+             output.extend(topics.iter().map(topics_to_bloom_filter));
+         }
+         output
+     }
+
+     /// Returns `true` if the bloom matches the topics
+     pub fn matches_topics(bloom: Bloom, topic_filters: &[BloomFilter]) -> bool {
+         if topic_filters.is_empty() {
+             return true;
+         }
+
+         // returns true if a filter matches
+         for filter in topic_filters.iter() {
+             let mut is_match = false;
+             for maybe_bloom in filter {
+                 is_match = maybe_bloom.as_ref().map(|b| bloom.contains_bloom(b)).unwrap_or(true);
```

Comment:
I think we should probably add most of the impls from here? https://docs.rs/ethbloom/0.13.0/src/ethbloom/lib.rs.html#174

cc @mattsse